File: apps/ios-native/WorkbenchApp/WebView/WebViewBridge.swift
import Foundation
import UIKit
import WebKit

/// WebView与原生功能的桥接类
/// 提供原生功能给H5调用
class WebViewBridge: NSObject {

    private weak var webView: WKWebView?
    private weak var viewController: UIViewController?
    private let bridgeCallback: BridgeCallback

    init(webView: WKWebView, viewController: UIViewController, bridgeCallback: BridgeCallback) {
        self.webView = webView
        self.viewController = viewController
        self.bridgeCallback = bridgeCallback
        super.init()
    }

    // MARK: - JavaScript Interface

    /**
     * 获取设备信息
     */
    @objc func getDeviceInfo() -> String {
        let deviceInfo = DeviceInfo(
            platform: "ios",
            version: UIDevice.current.systemVersion,
            model: UIDevice.current.model,
            brand: "Apple",
            screenWidth: Int(UIScreen.main.bounds.width),
            screenHeight: Int(UIScreen.main.bounds.height),
            density: UIScreen.main.scale
        )

        return deviceInfo.toJSONString()
    }

    /**
     * 获取网络状态
     */
    @objc func getNetworkStatus() -> String {
        let networkInfo = bridgeCallback.getNetworkStatus()
        return networkInfo.toJSONString()
    }

    /**
     * 调用相机
     */
    @objc func openCamera(_ callbackId: String) {
        bridgeCallback.openCamera { [weak self] result in
            DispatchQueue.main.async {
                self?.callJsCallback(callbackId: callbackId, result: result)
            }
        }
    }

    /**
     * 选择图片
     */
    @objc func pickImage(_ callbackId: String, maxCount: Int) {
        bridgeCallback.pickImage(maxCount: maxCount) { [weak self] result in
            DispatchQueue.main.async {
                self?.callJsCallback(callbackId: callbackId, result: result)
            }
        }
    }

    /**
     * 显示Toast消息
     */
    @objc func showToast(_ message: String, duration: String) {
        DispatchQueue.main.async {
            self.bridgeCallback.showToast(message: message, duration: duration)
        }
    }

    /**
     * 获取本地存储数据
     */
    @objc func getStorage(_ key: String) -> String {
        let value = bridgeCallback.getStorage(key: key)
        return value ?? "null"
    }

    /**
     * 设置本地存储数据
     */
    @objc func setStorage(_ key: String, value: String) {
        bridgeCallback.setStorage(key: key, value: value)
    }

    // MARK: - Private Methods

    private func callJsCallback(callbackId: String, result: Result<Any, Error>) {
        let response = BridgeResponse(
            callbackId: callbackId,
            success: result.isSuccess,
            data: result.success,
            error: result.failure?.localizedDescription
        )

        let jsCode = """
                if (window.NativeBridge && window.NativeBridge.callback) {
                    window.NativeBridge.callback('\(response.toJSONString())');
                }
            """

        webView?.evaluateJavaScript(jsCode, completionHandler: nil)
    }
}

// MARK: - Data Models

struct DeviceInfo: Codable {
    let platform: String
    let version: String
    let model: String
    let brand: String
    let screenWidth: Int
    let screenHeight: Int
    let density: CGFloat
}

struct NetworkInfo: Codable {
    let isConnected: Bool
    let type: String
    let strength: Int
}

struct BridgeResponse: Codable {
    let callbackId: String
    let success: Bool
    let data: Any?
    let error: String?

    enum CodingKeys: String, CodingKey {
        case callbackId, success, data, error
    }

    init(callbackId: String, success: Bool, data: Any?, error: String?) {
        self.callbackId = callbackId
        self.success = success
        self.data = data
        self.error = error
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(callbackId, forKey: .callbackId)
        try container.encode(success, forKey: .success)
        try container.encode(error, forKey: .error)

        if let data = data {
            if let stringData = data as? String {
                try container.encode(stringData, forKey: .data)
            } else if let intData = data as? Int {
                try container.encode(intData, forKey: .data)
            } else if let boolData = data as? Bool {
                try container.encode(boolData, forKey: .data)
            } else if let arrayData = data as? [String] {
                try container.encode(arrayData, forKey: .data)
            }
        }
    }
}

// MARK: - Bridge Callback Protocol

protocol BridgeCallback: AnyObject {
    func getNetworkStatus() -> NetworkInfo
    func openCamera(completion: @escaping (Result<String, Error>) -> Void)
    func pickImage(maxCount: Int, completion: @escaping (Result<[String], Error>) -> Void)
    func showToast(message: String, duration: String)
    func getStorage(key: String) -> String?
    func setStorage(key: String, value: String)
}

// MARK: - Extensions

extension Encodable {
    func toJSONString() -> String {
        do {
            let data = try JSONEncoder().encode(self)
            return String(data: data, encoding: .utf8) ?? "{}"
        } catch {
            return "{}"
        }
    }
}

extension Result {
    var isSuccess: Bool {
        switch self {
        case .success:
            return true
        case .failure:
            return false
        }
    }

    var success: Success? {
        switch self {
        case .success(let value):
            return value
        case .failure:
            return nil
        }
    }

    var failure: Failure? {
        switch self {
        case .success:
            return nil
        case .failure(let error):
            return error
        }
    }
}

File: apps/android-native/app/src/main/java/com/yourcompany/workbench/webview/WebViewBridge.kt
package com.yourcompany.workbench.webview

import android.content.Context
import android.webkit.JavascriptInterface
import android.webkit.WebView
import com.google.gson.Gson
import com.yourcompany.workbench.utils.Logger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * WebView与原生功能的桥接类
 * 提供原生功能给H5调用
 */
class WebViewBridge(
    private val context: Context,
    private val webView: WebView,
    private val bridgeCallback: BridgeCallback
) {
    private val gson = Gson()
    private val scope = CoroutineScope(Dispatchers.Main)

    /**
     * 获取设备信息
     */
    @JavascriptInterface
    fun getDeviceInfo(): String {
        return try {
            val deviceInfo = DeviceInfo(
                platform = "android",
                version = android.os.Build.VERSION.RELEASE,
                model = android.os.Build.MODEL,
                brand = android.os.Build.BRAND,
                screenWidth = context.resources.displayMetrics.widthPixels,
                screenHeight = context.resources.displayMetrics.heightPixels,
                density = context.resources.displayMetrics.density
            )
            gson.toJson(deviceInfo)
        } catch (e: Exception) {
            Logger.e("WebViewBridge", "获取设备信息失败", e)
            "{}"
        }
    }

    /**
     * 获取网络状态
     */
    @JavascriptInterface
    fun getNetworkStatus(): String {
        return try {
            val networkInfo = bridgeCallback.getNetworkStatus()
            gson.toJson(networkInfo)
        } catch (e: Exception) {
            Logger.e("WebViewBridge", "获取网络状态失败", e)
            "{}"
        }
    }

    /**
     * 调用相机
     */
    @JavascriptInterface
    fun openCamera(callbackId: String) {
        scope.launch {
            try {
                bridgeCallback.openCamera { result ->
                    val response = BridgeResponse(
                        callbackId = callbackId,
                        success = result.isSuccess,
                        data = result.getOrNull(),
                        error = result.exceptionOrNull()?.message
                    )
                    callJsCallback(response)
                }
            } catch (e: Exception) {
                Logger.e("WebViewBridge", "调用相机失败", e)
                val response = BridgeResponse(
                    callbackId = callbackId,
                    success = false,
                    error = e.message
                )
                callJsCallback(response)
            }
        }
    }

    /**
     * 选择图片
     */
    @JavascriptInterface
    fun pickImage(callbackId: String, maxCount: Int = 1) {
        scope.launch {
            try {
                bridgeCallback.pickImage(maxCount) { result ->
                    val response = BridgeResponse(
                        callbackId = callbackId,
                        success = result.isSuccess,
                        data = result.getOrNull(),
                        error = result.exceptionOrNull()?.message
                    )
                    callJsCallback(response)
                }
            } catch (e: Exception) {
                Logger.e("WebViewBridge", "选择图片失败", e)
                val response = BridgeResponse(
                    callbackId = callbackId,
                    success = false,
                    error = e.message
                )
                callJsCallback(response)
            }
        }
    }

    /**
     * 显示Toast消息
     */
    @JavascriptInterface
    fun showToast(message: String, duration: String = "short") {
        scope.launch {
            try {
                bridgeCallback.showToast(message, duration)
            } catch (e: Exception) {
                Logger.e("WebViewBridge", "显示Toast失败", e)
            }
        }
    }

    /**
     * 获取本地存储数据
     */
    @JavascriptInterface
    fun getStorage(key: String): String {
        return try {
            val value = bridgeCallback.getStorage(key)
            gson.toJson(value)
        } catch (e: Exception) {
            Logger.e("WebViewBridge", "获取存储数据失败", e)
            "null"
        }
    }

    /**
     * 设置本地存储数据
     */
    @JavascriptInterface
    fun setStorage(key: String, value: String) {
        scope.launch {
            try {
                bridgeCallback.setStorage(key, value)
            } catch (e: Exception) {
                Logger.e("WebViewBridge", "设置存储数据失败", e)
            }
        }
    }

    /**
     * 调用JS回调函数
     */
    private fun callJsCallback(response: BridgeResponse) {
        val jsCode = """
            if (window.NativeBridge && window.NativeBridge.callback) {
                window.NativeBridge.callback('${gson.toJson(response)}');
            }
        """.trimIndent()
        
        webView.post {
            webView.evaluateJavascript(jsCode, null)
        }
    }

    /**
     * 数据类定义
     */
    data class DeviceInfo(
        val platform: String,
        val version: String,
        val model: String,
        val brand: String,
        val screenWidth: Int,
        val screenHeight: Int,
        val density: Float
    )

    data class NetworkInfo(
        val isConnected: Boolean,
        val type: String,
        val strength: Int
    )

    data class BridgeResponse(
        val callbackId: String,
        val success: Boolean,
        val data: Any? = null,
        val error: String? = null
    )

    /**
     * 桥接回调接口
     */
    interface BridgeCallback {
        fun getNetworkStatus(): NetworkInfo
        fun openCamera(callback: (Result<String>) -> Unit)
        fun pickImage(maxCount: Int, callback: (Result<List<String>>) -> Unit)
        fun showToast(message: String, duration: String)
        fun getStorage(key: String): String?
        fun setStorage(key: String, value: String)
    }
}
